src/core/stdc/stdio.d

```diff
@@ -1146,97 +1146,30 @@ void setbuf(FILE* stream, char* buf);
 ///
 int  setvbuf(FILE* stream, char* buf, int mode, size_t size);
 
-version (MinGW)
-{
-    // Prefer the MinGW versions over the MSVC ones, as the latter don't handle
-    // reals at all.
-    ///
-    int __mingw_fprintf(FILE* stream, scope const char* format, ...);
-    ///
-    alias __mingw_fprintf fprintf;
-
-    ///
-    int __mingw_fscanf(FILE* stream, scope const char* format, ...);
-    ///
-    alias __mingw_fscanf fscanf;
-
-    ///
-    int __mingw_sprintf(scope char* s, scope const char* format, ...);
-    ///
-    alias __mingw_sprintf sprintf;
-
-    ///
-    int __mingw_sscanf(scope const char* s, scope const char* format, ...);
-    ///
-    alias __mingw_sscanf sscanf;
-
-    ///
-    int __mingw_vfprintf(FILE* stream, scope const char* format, va_list arg);
-    ///
-    alias __mingw_vfprintf vfprintf;
-
-    ///
-    int __mingw_vfscanf(FILE* stream, scope const char* format, va_list arg);
-    ///
-    alias __mingw_vfscanf vfscanf;
-
-    ///
-    int __mingw_vsprintf(scope char* s, scope const char* format, va_list arg);
-    ///
-    alias __mingw_vsprintf vsprintf;
-
-    ///
-    int __mingw_vsscanf(scope const char* s, scope const char* format, va_list arg);
-    ///
-    alias __mingw_vsscanf vsscanf;
-
-    ///
-    int __mingw_vprintf(scope const char* format, va_list arg);
-    ///
-    alias __mingw_vprintf vprintf;
-
-    ///
-    int __mingw_vscanf(scope const char* format, va_list arg);
-    ///
-    alias __mingw_vscanf vscanf;
-
-    ///
-    int __mingw_printf(scope const char* format, ...);
-    ///
-    alias __mingw_printf printf;
-
-    ///
-    int __mingw_scanf(scope const char* format, ...);
-    ///
-    alias __mingw_scanf scanf;
-}
-else
-{
-    ///
-    int fprintf(FILE* stream, scope const char* format, ...);
-    ///
-    int fscanf(FILE* stream, scope const char* format, ...);
-    ///
-    int sprintf(scope char* s, scope const char* format, ...);
-    ///
-    int sscanf(scope const char* s, scope const char* format, ...);
-    ///
-    int vfprintf(FILE* stream, scope const char* format, va_list arg);
-    ///
-    int vfscanf(FILE* stream, scope const char* format, va_list arg);
-    ///
-    int vsprintf(scope char* s, scope const char* format, va_list arg);
-    ///
-    int vsscanf(scope const char* s, scope const char* format, va_list arg);
-    ///
-    int vprintf(scope const char* format, va_list arg);
-    ///
-    int vscanf(scope const char* format, va_list arg);
-    ///
-    int printf(scope const char* format, ...);
-    ///
-    int scanf(scope const char* format, ...);
-}
+///
+int fprintf(FILE* stream, scope const char* format, ...);
```

Comment:
Have you verified that those work with 80-bit `real`? Last time I checked, these are the MSVC functions, where `long double` is 64-bit, so that `printf("%Lf", myReal)` prints garbage.

Reply:
No really, I only checked the unittests which seem to be OK but might not test this at all. How do LDC/DMD handle this when targeting MSVC? Do you use 64bit reals instead? We could do that in GDC, just set `-mlong-double-64` by default.

To be honest, MinGW is a low-priority target for me. It's not well supported, developing for windows is frustrating (even running the GDC testsuite is a hacky process, as it uses symlinks..., there's no well-working SSH server, IO and building GCC is way too slow, ...). People will also always find something to complain about (uses emulated TLS, uses dwarf debug info, c++ abi is not compatible with MSVC, use .a import libraries) so I want to keep as little MinGW specific code as possible. Preferably, windows users should just use LDC, but we also need something to compile GDC with the D frontend for windows hosts and I'm not sure if GCC can easily be built with clang.

Reply:
> How do LDC/DMD handle this when targeting MSVC? Do you use 64bit reals instead?

LDC, unlike DMD, uses 64-bit `real`, but for MSVC targets only, i.e., (only partially supported) MinGW targets use 80-bit extended precision.

> People will also always find something to complain about (uses emulated TLS, uses dwarf debug info, c++ abi is not compatible with MSVC, use .a import libraries) so I want to keep as little MinGW specific code as possible.

Same here. :]

> Preferably, windows users should just use LDC

Heh, I always thought MinGW users should just use GDC. ;)

Reply:
The real irony here is that as DMD is moving away from DMC/optlink they are considering lld+mingw libraries.

Reply:
Oh, I didn't know LDC supports a MinGW target. So I'll revert these changes here.

> Heh, I always thought MinGW users should just use GDC. ;)
This makes me wonder: Do you actually use native TLS when targeting MinGW or the libgcc emutls? If you use emutls, how did you manage to integrate it with the GC? I think in the history of GDC, we never really had reliable MinGW builds because of that, the GC has always been broken. The only exception might be Daniels old builds which patched GCC and binutils to support native windows TLS ;-)

Reply:
@kinke Do you need any of the other MinGW definitions here? `fenv.d`, `math.d` and the rest of the changed files?

Reply:
@thewilsonator 
> The real irony here is that as DMD is moving away from DMC/optlink they are considering lld+mingw libraries.

I think when using LLD+MinGW import libraries only you're not affected by most of the problems: DWARF debug info / object file size / missing TLS support / C++ ABI are GCC and binutils problems, using LLD/clang you're probably fine.

Reply:
Ahh, good to know.

Reply:
> Oh, I didn't know LDC supports a MinGW target. So I'll revert these changes here.

LDC used to support MinGW (but not 100% I guess - especially wrt. C++ ABI/name-mangling), but doesn't anymore - MSVC is our primary/exclusive Windows target. There shouldn't be that many changes to get it working again (mainly druntime/Phobos), but it's been like that for years now, so I don't think proper MinGW support is coming back anytime soon (I have 0 interest in it).

Wrt. TLS, according to https://run.dlang.io/is/PR7PuP we use native TLS. clang has the option to produce MSVC objects with dwarf debuginfo; apparently some Windows guys prefer gdb over the Visual Studio debugger. I'd also expect LLD to be usable as linker for MinGW in general (support for .a, dwarf, huge objects).

Reply:
> LDC used to support MinGW (but not 100% I guess - especially wrt. C++ ABI/name-mangling & EH)

Yes, LDC/MinGW was the fully working first non-DMD compiler targeting Windows. The C++ ABI/EH interface wasn't a thing back then.

Reply:
OK, reverted this change.